Support proof closures

jaybosamiya-ms opened the pull request:
See https://github.com/verus-lang/verus/pull/1524

parno commented:
I might be missing something, but I don't see support for the various proof function options, i.e., for something like this example from Chris's PR:
```
        proof fn testfn() {
            let tracked f = proof_fn[Mut, Copy, Send, Sync]|y: u64| -> (z: u64)
                requires y == 2
                ensures z == 2
            {
                y
            };
            helper(f);
        }
```

jaybosamiya-ms commented:
Good catch! Indeed that is missing, will add soon.

jaybosamiya-ms commented:
Added support for the proof function options `[Mut, ...]` (calling them characteristics, following what @Chris-Hawblitzel wrote), including the `ReqEns<...>`. I also added the type-level support things.

parno reviewed (approved):
Looks good to me.  I cribbed off these changes for the verus-analyzer updates :)
